### ISSUE TITLE
Class definitions are not Hoisted, promoting race conditions and undefined classes

### ISSUE THREAD

**Issue by aodl** [Microsoft/TypeScript#15468]
**TypeScript Version:**  whichever version is currently running in TypeScript Playground (2.3 I think)

**Code**
```ts
let myObject = new MyClass();
// MyClass definition goes here
```
**Expected behavior:**
No runtime error (or at least a compile time error to warn about the runtime error)

**Actual behavior:**
Runtime error "Object doesn't support this action" (in Edge) or "... is not a constructor" (in Chrome)

**Comment**
It is apparently valid at compile time to declare/define a class at the bottom of a file, and instantiate and use this object somewhere else in that file (as is the case in C#, Java, etc.).

However, in the transcompiled javascript the 'class' definition is not hoisted to the top of the file, or above where it is used (this is a problem as the JavaScript equivalent of declaring a TypeScript class is to run an immediately invoked function that returns a constructor, but this must happen before instantiation).

Moving class declarations above where the classes are instantiated solves this problem (TypeScript does not warn or prompt about this). Due to this, it is easy to introduce a race condition, such as the one below.
```
$(() => {
    let myObject = new MyClass();
});
// MyClass definition goes here
```

**Comment by fatcerberus** [comment]
Classes are not hoisted:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Hoisting
